File: src/commands/summon.py
import asyncio
from datetime import datetime

from telegram import ChatMember, InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.constants import KeyboardButtonStyle, ParseMode
from telegram.ext import CallbackContext, ContextTypes

import commands
import utils
from config.db import get_db
from utils.decorators import command
from utils.messages import get_message

summon_log = {}


async def perform_summon(
    context: ContextTypes.DEFAULT_TYPE,
    chat_id: int,
    group_id: int,
    group_name: str,
):
    async with get_db() as conn:
        async with conn.execute(
            """
            SELECT user_id, chat_id
            FROM summon_group_members
                JOIN summon_groups ON summon_groups.id = summon_group_members.group_id
            WHERE group_id = ?;
            """,
            (group_id,),
        ) as cursor:
            group_members = [(row[0], row[1]) for row in await cursor.fetchall()]

    async def check_member(user_id: int, member_chat_id: int) -> str | None:
        try:
            member = await context.bot.get_chat_member(member_chat_id, user_id)
            if member.status not in [ChatMember.LEFT, ChatMember.BANNED]:
                username = await utils.get_username(user_id, context)
                if username:
                    return f"@{username}"
        except Exception as e:
            print(f"Error getting member {user_id}: {e}")
        return None

    members = [
        member
        for member in await asyncio.gather(
            *(check_member(user_id, member_chat_id) for user_id, member_chat_id in group_members)
        )
        if member
    ]
    keyboard = InlineKeyboardMarkup(
        [
            [
                InlineKeyboardButton(
                    "✅ Join",
                    callback_data=f"sg:join,{group_id}",
                    style=KeyboardButtonStyle.SUCCESS,
                ),
                InlineKeyboardButton(
                    "❌ Leave",
                    callback_data=f"sg:leave,{group_id}",
                    style=KeyboardButtonStyle.DANGER,
                ),
            ],
            [
                InlineKeyboardButton(
                    "🔁 Resummon",
                    callback_data=f"sg:resummon,{group_id}",
                    style=KeyboardButtonStyle.PRIMARY,
                )
            ],
        ]
    )

    if not members:
        await context.bot.send_message(
            chat_id,
            f"No users in group '{group_name}'.",
            reply_markup=keyboard,
            parse_mode=ParseMode.HTML,
        )
        summon_log[group_id] = datetime.now()
        return

    for index in range(0, len(members), 5):
        chunk = members[index : index + 5]
        await context.bot.send_message(
            chat_id,
            f"[{group_name}] ({len(members)} members) {' '.join(chunk)}",
            parse_mode=ParseMode.HTML,
            reply_markup=keyboard if index + 5 >= len(members) else None,
        )

    summon_log[group_id] = datetime.now()


async def summon_keyboard_button(update: Update, context: CallbackContext) -> None:
    message = get_message(update)

    if not message:
        return
    query = update.callback_query
    if not query or not query.data or not query.from_user or not query.message:
        return

    action, group_id_str = query.data.replace("sg:", "").split(",")
    group_id = int(group_id_str)

    async with get_db() as conn:
        if action == "join":
            await conn.execute(
                """
                INSERT INTO summon_group_members (group_id, user_id)
                VALUES (?, ?)
                ON CONFLICT (group_id, user_id) DO NOTHING
                """,
                (group_id, query.from_user.id),
            )
            await query.answer("Joined group.")
            return
        elif action == "leave":
            await conn.execute(
                "DELETE FROM summon_group_members WHERE group_id = ? AND user_id = ?",
                (group_id, query.from_user.id),
            )
            await query.answer("Left group.")
            return
        elif action == "resummon":
            last_tag_time = summon_log.get(group_id)
            if last_tag_time and (datetime.now() - last_tag_time).seconds < 60:
                remaining_seconds = 60 - (datetime.now() - last_tag_time).seconds
                await query.answer(
                    f"You can only resummon once every 60 seconds. Wait {remaining_seconds}s."
                )
                return

            await query.answer("Resummoning...")
            async with get_db() as conn:
                async with conn.execute(
                    "SELECT group_name FROM summon_groups WHERE id = ?", (group_id,)
                ) as cursor:
                    result = await cursor.fetchone()
                    group_name = result[0] if result else "Unknown"

            await perform_summon(context, message.chat_id, group_id, group_name)
            return


@command(
    triggers=["summon"],
    usage="/summon [GROUP_NAME]",
    example="/summon SwitchPlayers",
    description="Tag users present in a group of tags. Join by using keyboard buttons.",
)
async def summon(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = get_message(update)
    if not message:
        return
    if not update.effective_chat or not update.effective_user:
        return

    if update.effective_chat.type == "private":
        await message.reply_text("This command can only be used in group chats.")
        return

    if not context.args:
        await commands.usage_string(message, summon)
        return

    group_name = context.args[0].lower()

    async with get_db() as conn:
        async with conn.execute(
            "SELECT id FROM summon_groups WHERE group_name = ? COLLATE NOCASE AND chat_id = ?",
            (group_name, update.effective_chat.id),
        ) as cursor:
            result = await cursor.fetchone()

        if result:
            group_id = result[0]
        else:
            cursor = await conn.execute(
                "INSERT INTO summon_groups (group_name, chat_id, creator_id) VALUES (?, ?, ?)",
                (group_name, update.effective_chat.id, update.effective_user.id),
            )
            group_id = cursor.lastrowid
            if group_id is None:
                raise RuntimeError("Failed to create summon group.")
            await conn.execute(
                "INSERT INTO summon_group_members (group_id, user_id) VALUES (?, ?)",
                (group_id, update.effective_user.id),
            )

    await perform_summon(context, update.effective_chat.id, group_id, group_name)

File: src/features/summon.ts
import {
  answerCallback,
  callbackData,
  Effect,
  getChatMember,
  html,
  Schema,
  sendMessage,
} from "telly";

import {
  answer,
  type CommandDefinition,
  usage,
} from "../app/command.ts";
import { callbackRoute } from "../app/callback.ts";
import { rowNumber, rowString } from "../app/database.ts";
import type { AppDependencies } from "../app/dependencies.ts";

const SummonCallback = callbackData("summon", Schema.Struct({
  action: Schema.Literals(["join", "leave", "resummon"]),
  groupId: Schema.Int,
}));

function summonKeyboard(groupId: number) {
  return {
    inlineKeyboard: [
      [
        { ...SummonCallback.button("✅ Join", { action: "join", groupId }), style: "success" as const },
        { ...SummonCallback.button("❌ Leave", { action: "leave", groupId }), style: "danger" as const },
      ],
      [{ ...SummonCallback.button("🔁 Resummon", { action: "resummon", groupId }), style: "primary" as const }],
    ],
  };
}

function performSummon(
  dependencies: AppDependencies,
  chatId: number,
  groupId: number,
  groupName: string,
) {
  return Effect.gen(function* () {
    const rows = yield* dependencies.database.all(
      `SELECT members.user_id
       FROM summon_group_members members
       JOIN summon_groups groups ON groups.id = members.group_id
       WHERE members.group_id = ? AND groups.chat_id = ?`,
      [groupId, chatId],
    );
    const members = yield* Effect.forEach(rows, (row) =>
      Effect.result(getChatMember({ chatId, userId: rowNumber(row, "user_id") })).pipe(
        Effect.map((result) => result._tag === "Failure" || ["left", "kicked"].includes(
          result.success.status,
        )
          ? undefined
          : `@${result.success.user.username ?? result.success.user.firstName}`),
      ), { concurrency: 8 });
    const active = members.filter((member): member is string => member !== undefined);
    if (active.length === 0) {
      yield* sendMessage({
        chatId,
        parseMode: "HTML",
        replyMarkup: summonKeyboard(groupId),
        text: `No users in group '${html.escape(groupName)}'.`,
      });
      return;
    }
    for (let index = 0; index < active.length; index += 5) {
      const chunk = active.slice(index, index + 5);
      yield* sendMessage({
        chatId,
        parseMode: "HTML",
        ...(index + 5 >= active.length ? { replyMarkup: summonKeyboard(groupId) } : {}),
        text: `[${html.escape(groupName)}] (${active.length} members) ${chunk.map(html.escape).join(" ")}`,
      });
    }
  });
}

export function summonFeature(dependencies: AppDependencies) {
  const lastSummon = new Map<number, number>();
  const definition: CommandDefinition = {
    description: "Tag members of a named group and let people join from buttons.",
    example: "/summon SwitchPlayers",
    names: ["summon"],
    run: Effect.fn("summon")(function* (match) {
      const user = match.message.from;
      if (match.message.chat.type === "private") {
        return yield* answer(match.message, "This command can only be used in group chats.");
      }
      const groupName = match.args[0]?.toLowerCase();
      if (user === undefined || groupName === undefined) {
        return yield* usage(match.message, definition);
      }
      const inserted = yield* dependencies.database.execute(
        `INSERT OR IGNORE INTO summon_groups (group_name, chat_id, creator_id)
         VALUES (?, ?, ?)`,
        [groupName, match.message.chat.id, user.id],
      );
      const row = yield* dependencies.database.one(
        `SELECT id FROM summon_groups
         WHERE group_name = ? COLLATE NOCASE AND chat_id = ?`,
        [groupName, match.message.chat.id],
      );
      if (row === undefined) return yield* Effect.die(new Error("Summon group insert failed"));
      const groupId = rowNumber(row, "id");
      if (inserted.rowsAffected > 0) {
        yield* dependencies.database.execute(
          "INSERT OR IGNORE INTO summon_group_members (group_id, user_id) VALUES (?, ?)",
          [groupId, user.id],
        );
      }
      yield* performSummon(dependencies, match.message.chat.id, groupId, groupName);
      lastSummon.set(groupId, dependencies.monotonicMilliseconds());
    }),
    usage: "/summon [group name]",
  };
  const callback = callbackRoute("summon", SummonCallback, Effect.fn("summonCallback")(function* ({
    callbackQuery,
    data,
  }) {
    if (data.action === "join") {
      yield* dependencies.database.execute(
        "INSERT OR IGNORE INTO summon_group_members (group_id, user_id) VALUES (?, ?)",
        [data.groupId, callbackQuery.from.id],
      );
      yield* answerCallback(callbackQuery, { text: "Joined group." });
      return;
    }
    if (data.action === "leave") {
      yield* dependencies.database.execute(
        "DELETE FROM summon_group_members WHERE group_id = ? AND user_id = ?",
        [data.groupId, callbackQuery.from.id],
      );
      yield* answerCallback(callbackQuery, { text: "Left group." });
      return;
    }
    const last = lastSummon.get(data.groupId);
    const elapsed = last === undefined
      ? Number.POSITIVE_INFINITY
      : dependencies.monotonicMilliseconds() - last;
    if (elapsed < 60_000) {
      yield* answerCallback(callbackQuery, {
        text: `You can only resummon once every 60 seconds. Wait ${Math.ceil((60_000 - elapsed) / 1_000)}s.`,
      });
      return;
    }
    const group = yield* dependencies.database.one(
      "SELECT group_name, chat_id FROM summon_groups WHERE id = ?",
      [data.groupId],
    );
    if (group === undefined) {
      yield* answerCallback(callbackQuery, { text: "Summon group no longer exists." });
      return;
    }
    yield* answerCallback(callbackQuery, { text: "Resummoning..." });
    yield* performSummon(
      dependencies,
      rowNumber(group, "chat_id"),
      data.groupId,
      rowString(group, "group_name"),
    );
    lastSummon.set(data.groupId, dependencies.monotonicMilliseconds());
  }));
  return { callback, commands: [definition] as const };
}
